tensorpack/graph_builder/distributed.py

```diff
@@ -319,21 +316,21 @@ def build(self, get_grad_fn, get_opt_fn):
                         self.gpu_shadow_vars[k].append((grad_aggregation_variable, None))
             assert len(self.gpu_shadow_vars) == len(self.grad_list)
 
-            # Apply new gradients to variables.
-            aggregation_ops_list = []
-            for k in range(len(self.grad_list)):
-                for idx, (grad, var) in enumerate(self.grad_list[k]):
-                    with tf.device(self.raw_devices[k]), tf.variable_scope("aggregation_variables", reuse=True):
+            # Apply new gradients to aggregation variables.
+            for k, (grad_list, device) in enumerate(zip(self.grad_list, self.raw_devices)):
+                for idx, (grad, var) in enumerate(grad_list):
+                    with tf.device(device), tf.variable_scope("aggregation_variables", reuse=True):
                         grad_aggregation_variable_name = self._get_grad_aggregator_variable_name(
                             tower_id=k, variable_id=idx)
                         grad_aggregator = tf.get_variable(grad_aggregation_variable_name)
                         aggregation_ops_list.append(grad_aggregator.assign_add(grad))
                         self.gpu_shadow_vars[k][idx] = (self.gpu_shadow_vars[k][idx][0], var)
-            aggregation_ops = tf.group(*aggregation_ops_list)
-        else:
-            aggregation_ops = tf.group()
+        aggregation_ops = tf.group(*aggregation_ops_list)
```

Comment:
I am using a `tf.group()` here so that it matches the style of the tensorpack code, can drop since it's possible to just pass a list into `control_dependencies`

Reply:
I'll advocate for the simpler code as a general principle, but if it looks like other stuff here, then either way is fine.